iohub/dataset_base.py

```diff
@@ -0,0 +1,59 @@
+from abc import ABC, abstractmethod
+from typing import Any, Generator
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+
+class BaseFOV(ABC):
+    # NOTE: not using the `data` method from ngff Position
+
+    @property
+    @abstractmethod
+    def channels(self) -> list[str]:
+        raise NotImplementedError
+
+    def channel_index(self, key: str) -> int:
+        """Return index of given channel."""
+        return self.channels.index(key)
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, int, int, int, int]:
+        # NOTE: suggestion, fix dimension to 5?
+        # We could me more restrictive than ome-zarr
+        raise NotImplementedError
+
+    @abstractmethod
+    def __getitem__(self, key: Any) -> ArrayLike:
```

Comment:
I was a bit more strict. I added the types supported by zarr. What do you think?

Reply:
What doe ellipsis in `Union` mean?

Reply:
The ellipsis is for the tuple, the syntax is a bit misleading.
With python 3.11 it can be typed as, which is clearer in my opinion:

```python3
int | slice | tuple[int | slice, ...]
```

It can be an int, a slice, or a tuple with ints or slices.

Reply:
Oops wrong bracket count. Thanks for the explanation!